Python çalışma ortamına veri okuma yapmadan önce, çalışma klasörümüzde öncelikle hangi dosyaların bulunduğuna ilişkin fikrimizin olması gerekir. os modülünü yükleyerek aşağıdaki kodları çalıştırdığımızda, çalışma klasörümüzde yüklü olan dosyalar listelenecektir. 


In [1]:
import os
yuklu_dosyalar = os.listdir('.')
print(yuklu_dosyalar)

['.ipynb_checkpoints', 'cars_train_annos.mat', 'coor.csv', 'credit.txt', 'EmployeeData.json', 'istanbul_hastane.json', 'lorem_ipsum.txt', 'MERS_IGS.csv', 'mutluluk_indeksi.xlsx', 'netflix.csv', 'sample.json', 'sample.txt', 'SASVisualForecasting_sampledatasets.zip', 'tax.dta', 'tax.sas7bdat', 'titanic.csv', 'traffic_density_20200711.csv', 'traffic_density_202010.csv', 'unite 09 veri okuma.ipynb']


## 1. Düz Dosyaları Okuma

.txt ya da .csv uzantılı dosyalar düz dosyalardır. .txt dosyaları metin dosyalarıdır. .csv dosyaları ise (Comma Separated Values), virgül ile ayrılmış, satır ve sütunlardan oluşan değerleri barındıran düz dosya tipleridir. 

Çalışma klasörümüzde bulunan lorem_ipsum.txt adındaki dosyayı salt okunur olarak Python çalışma ortamımıza ekleyelim. Lorem Ipsum, dizgi ve baskı endüstrisinde kullanılan Latince mıgır metinlerdir.

In [2]:
# Bir dosya açalım: dosya
dosya = open('lorem_ipsum.txt', "r") # Salt okunur olarak eklemek istediğimiz için "r" kullandık. 

# Dosyayı üzerine yazılabilir formatta açabilmek için "w" şeklinde bir yazım tercih etmemiz gerekir. 

In [3]:
# Dosyadaki metni ekrana yazdıralım.
print(dosya.read())

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc volutpat vestibulum turpis vitae eleifend. Nullam vitae sollicitudin erat, non interdum felis. Aenean fringilla nec orci eu eleifend. Suspendisse semper turpis nec libero finibus, eget viverra ante molestie. Proin quis pulvinar metus. Nunc tortor felis, faucibus ac purus eu, suscipit dictum dolor. Nullam vel mi sed nibh varius cursus.

Phasellus aliquet nunc nec ex convallis tincidunt. Phasellus ac pretium massa. Aenean purus urna, egestas sit amet pharetra non, volutpat ut metus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi non mauris non mauris interdum eleifend. Cras ex purus, dignissim sit amet velit at, molestie scelerisque justo. Phasellus sit amet placerat lorem. Pellentesque sodales, lacus in cursus faucibus, nulla sem egestas nisl, sit amet molestie lorem nunc eget nibh. Vestibulum fringilla risus nec ipsum semper, fermentum facilisis eros aliquam. Nulla porttitor risus ut diam accumsan maximus at 

In [4]:
# Dosyanın açık mı kapalı mı olduğunu denetleyelim.
print(dosya.closed)

False


In [5]:
#Dosyamız açık olduğu için kapatalım.
dosya.close()

In [6]:
# Tekrar kontrol edelim.
print(dosya.closed)

True


.csv uzantılı dosyaları (Comma Separated Values) listelerden oluşan verileri barındırmakta olan düz bir dosya türüdür. Genellikle farklı uygulamalar arasında veri değişimini sağlamak amacıyla sıklıkla kullanılır. Veriler; virgül, noktalı virgül, aralık ya da başka işaretler kullanılarak ayrılır (delimiter). .csv uzantılı dosyaları metin editörleri (notepad vb.) ya da hesap tablosu programları (excel vb.) ile okuyabilmek ya da düzenleyebilmek olanaklıdır. .csv dosyalarının ilk satırı (header), verinin içindeki sütunların isimlerini barındırır.

NumPy kütüphanesi kullanarak düz dosyaları Python ortamına akratabilmekteyiz. Aşağıdaki örnekte, 1000 satırdan oluşan ve sadece enlem-boylam verilerini içeren bir .csv dosyasını açmaktayız. Bu dosyadaki veriler, virgül ile ayrıldığı için `delimiter=","` ifadesini kullandık. Bu tarz gerçekçi ama uydurma veriler üretebilmek için https://www.mockaroo.com/ adresini kullanabilirsiniz. 

In [7]:
import numpy as np
dosya_ismi = "coor.csv"
veri = np.loadtxt(dosya_ismi, delimiter = ',')
print(veri)

[[-11.1215276 -75.3587075]
 [  3.4802964  97.8120924]
 [  8.2301756 -80.5549561]
 ...
 [ 55.7213148  37.6243984]
 [ 27.569517  110.001922 ]
 [ 38.1788095  15.4976181]]


Bir sonraki örneğimizde kredi kartı numaralarını ve yıllık harcamalarını içeren iki sütuna sahip bir .txt dosyasını NumPy kütüphanesi yardımıyla okuyalım. .txt dosyasını okumak için NumPy kütüphanesindeki `.loadtxt()` yöntemini kullanacağınız. Parantez içinde dosya ismini girdikten sonra, bu dosyadaki ayracın nokta virgül olduğunu `delimiter = ";"` ifadesi ile göstereceğiz. İlk satırı okumak istemediğimiz için `skiprow = 1` ifadesini kullanıyoruz. Sadece kredi kartı limitlerini gösteren ikinci sütunu (birinci indeksi) data dosyası içerisine dahil etmek istediğimiz için de, `usecols=[1]` ifadesini argüman olarak ekliyoruz.

In [8]:
import numpy as np

# Dosya ismini belirleyelim: dosya
dosya = 'credit.txt'

# Veri olarak yükleme: data
data = np.loadtxt(dosya, delimiter=';', skiprows=1, usecols=[1])

# Ekrana verelim.
print(data)

[40017. 11854. 23624. 24944. 24996. 32537. 33251. 12867.  8936. 32053.
 37373. 27219. 44527.  8242. 34053.  9322. 11661. 28221.  6211. 26027.
 14473. 10654. 13069. 25816. 15610. 34821. 40970. 28233. 13378. 37949.
 31140. 25712. 39802. 36141. 43455. 13971. 40399. 34560. 37649. 13915.
 26944. 18319. 24656. 15668. 14886. 30706. 44958. 48397. 30186. 43141.
 25188. 26508.  9025. 25843. 43032.  8230. 46102. 11300. 17730. 26016.
 30092.  5999. 48292. 30295. 25528. 30809. 36954. 35716. 17544. 28282.
 28464.  9946.  7287. 49619. 42702. 14838. 40901. 32313. 11183. 42414.
 22721. 17835. 34900. 28810. 38037. 30236. 33060.  6652. 30364. 24732.
 18577. 49337. 16455. 35248. 39711. 23480. 25204. 31865. 28664.]


Çoğunlukla farklı veri tiplerinden oluşan veri setleri üzerine çalışırız. Bir sütun metinlerden oluşurken, diğer bir sütun ondalık sayılardan oluşabilir. np.loadtxt() fonksiyonu, metin harici veri tiplerini yüklemekte başarısız olacaktır. Farklı veri tiplerini yönetebilmek için `np.genfromtxt()` fonksiyonu kullanılabilir. Veri tipini özel olarak belirtmezsek `(dtype = None)`, bu fonksiyon otomatik olarak sütunların hangi veri tipini içerdiğini bulacaktır. Çalışma klasörümüzde bulunan ve Titanik gemisi yolcularına ilişkin verileri barındıran titanic.csv dosyasını yüklemeye çalışalım. 

In [9]:
data = np.genfromtxt('titanic.csv', delimiter=',', names=True, dtype=None)
print(data[0:10]) # Verinin ilk on sütunu ekrana verelim.
print(np.shape(data))

[( 1, 0, 3, b'male', 22., 1, 0, b'A/5 21171',  7.25  , b'', b'S')
 ( 2, 1, 1, b'female', 38., 1, 0, b'PC 17599', 71.2833, b'C85', b'C')
 ( 3, 1, 3, b'female', 26., 0, 0, b'STON/O2. 3101282',  7.925 , b'', b'S')
 ( 4, 1, 1, b'female', 35., 1, 0, b'113803', 53.1   , b'C123', b'S')
 ( 5, 0, 3, b'male', 35., 0, 0, b'373450',  8.05  , b'', b'S')
 ( 6, 0, 3, b'male', nan, 0, 0, b'330877',  8.4583, b'', b'Q')
 ( 7, 0, 1, b'male', 54., 0, 0, b'17463', 51.8625, b'E46', b'S')
 ( 8, 0, 3, b'male',  2., 3, 1, b'349909', 21.075 , b'', b'S')
 ( 9, 1, 3, b'female', 27., 0, 2, b'347742', 11.1333, b'', b'S')
 (10, 1, 2, b'female', 14., 1, 0, b'237736', 30.0708, b'', b'C')]
(891,)


C:\Users\ACER\anaconda3\lib\site-packages\ipykernel_launcher.py:1: VisibleDeprecationWarning: Reading unicode strings without specifying the encoding argument is deprecated. Set the encoding, use None for the system default.
  """Entry point for launching an IPython kernel.


`np.genfromtxt()` fonksiyonunda kullanılan ilk argüman dosya ismini belirtmektedir. İkinci argüman ise ayraç olan virgülü belirttiğimiz `delimiter` parametresidir. Üçüncü  `names` argümanı ise, dosyanın bir başlık satırı (header) içerdiğini belirtmek için kullanılmaktadır. Dosyamız bir header içerdiği için True parametresi kullanılmıştır. Dosyamızın boyutunu sorguladığımızda sadece 891 satırdan oluştuğunu görmekteyiz. Oysa ki bu dosya 891 satır, 11 sütundan oluşmaktadır. Bunun NumPy dizilerinin aynı tipte verileri barındırması zorunda olmasıdır. Dolayısıyla `np.genfromtxt()` fonksiyonu ile yüklenen dosyamız, her bir satırın bir dizi olduğu tek boyutlu bir liste olarak karşımıza çıkmaktadır. 

`np.genfromtxt()` fonksiyonuna alternatif olarak `np.recfromcsv()` fonksiyonu bulunmaktadır. İki fonksiyon da aynı davranış biçimine sahip olsa da, `np.recfromcsv()` fonksiyonunun sabit veri tipi None olmaktadır (dtype = None). 


In [10]:
# Dosya ismini belirleyelim: 
dosya = 'titanic.csv'

# np.recfromcsv() dosyasını kullanarak verimizi yükleyelim.
data = np.recfromcsv(dosya)

# data verisinin ilk 10 elemanını ekrana verelim.
print(data[:10])

# data verisinin boyutunu inceleyelim.
print(np.shape(data))

[( 1, 0, 3, b'male', 22., 1, 0, b'A/5 21171',  7.25  , b'', b'S')
 ( 2, 1, 1, b'female', 38., 1, 0, b'PC 17599', 71.2833, b'C85', b'C')
 ( 3, 1, 3, b'female', 26., 0, 0, b'STON/O2. 3101282',  7.925 , b'', b'S')
 ( 4, 1, 1, b'female', 35., 1, 0, b'113803', 53.1   , b'C123', b'S')
 ( 5, 0, 3, b'male', 35., 0, 0, b'373450',  8.05  , b'', b'S')
 ( 6, 0, 3, b'male', nan, 0, 0, b'330877',  8.4583, b'', b'Q')
 ( 7, 0, 1, b'male', 54., 0, 0, b'17463', 51.8625, b'E46', b'S')
 ( 8, 0, 3, b'male',  2., 3, 1, b'349909', 21.075 , b'', b'S')
 ( 9, 1, 3, b'female', 27., 0, 2, b'347742', 11.1333, b'', b'S')
 (10, 1, 2, b'female', 14., 1, 0, b'237736', 30.0708, b'', b'C')]
(891,)


C:\Users\ACER\anaconda3\lib\site-packages\numpy\lib\npyio.py:2372: VisibleDeprecationWarning: Reading unicode strings without specifying the encoding argument is deprecated. Set the encoding, use None for the system default.
  output = genfromtxt(fname, **kwargs)


Bu şekilde bir listelerden ya da dizilerden oluşan liste kullanımı, veri bilimi açısından çeşitli eksiklikler barındırmaktadır. Bir veri bilimci, aşağıdaki özelliklere sahip veriler üzerinde çalışmayı her zaman tercih eder:

1) İki boyutlu ve anlamlı bir şekilde etiketlenmiş veriler (Hangi sütunun ya da satırın neyi ifade ettiği bilinmeli).

2) Sütunlarda genellikle farklı tiplerde veriler bulunmalı.

3) Veriler birleştirilebilir, dilimlenebilir, gruplanabilir, değiştirilebilir olmalı.

4) Üzerinde istatistiki analizler yapılabilmeli.

5) Zaman serileri olarak çalışabilmeli.

Tüm bu gereksinimleri karşılayabilmek için veri çerçeveleri (DataFrame) ve bunları yönetebileceğimiz pandas kütüphanesini kullanmak daha uygun olacaktır. Bir veri çerçevesi, farklı veri tiplerindeki veri değerlerini içeren, sıralı sütunlardan oluşan ve iki indeksli (hem satırlar hem sütunlar için indekslenmiş) veri yapılarıdır.

İlk satırı sütün isimlerinden oluşan bir .csv dosyasını (bu örnekte de Titanik veri setini kullanalım) pandas kütüphanesi ile yüklemek için önce bu kütüphaneyi yüklemek gerekir. Aşağıdaki kodlar yardımıyla bu veri çerçevesini okuyalım.

In [11]:
# pandas kütüphanesini yükleyelim. pd olarak kısaltalım.
import pandas as pd

# dosya ismini atayalım.
dosya1 = 'titanic.csv'

# Veri çerçevesini (DataFrame) oluşturalım. pd.read_csv(dosya) yöntemi ile csv dosyaları okunur.
df_titanic = pd.read_csv(dosya1)

# Veri çerçevesinin ilk beş satırını .head() yöntemi ile ekrana verelim
print(df_titanic.head())

   PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0            1         0       3    male  22.0      1      0   
1            2         1       1  female  38.0      1      0   
2            3         1       3  female  26.0      0      0   
3            4         1       1  female  35.0      1      0   
4            5         0       3    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S  


Aynı işlemi Netflix'te gösterimde olan yapımları içeren veri seti üzerinde de gerçekleştirelim.

In [12]:
# pandas kütüphanesini yükleyelim. pd olarak kısaltalım.
import pandas as pd

# dosya ismini atayalım.
dosya2 = 'netflix.csv'

# Veri çerçevesini (DataFrame) oluşturalım. pd.read_csv(dosya) yöntemi ile csv dosyaları okunur.
df_netflix = pd.read_csv(dosya2)

# Veri çerçevesinin ilk beş satırını .head() yöntemi ile ekrana verelim
print(df_netflix.head())

         id      tur                                   baslik  \
0  81145628    Movie  Norm of the North: King Sized Adventure   
1  80117401    Movie               Jandino: Whatever it Takes   
2  70234439  TV Show                       Transformers Prime   
3  80058654  TV Show         Transformers: Robots in Disguise   
4  80125979    Movie                             #realityhigh   

                   yonetmen  \
0  Richard Finn, Tim Maltby   
1                       NaN   
2                       NaN   
3                       NaN   
4          Fernando Lebrija   

                                                cast  \
0  Alan Marriott, Andrew Toth, Brian Dobson, Cole...   
1                                   Jandino Asporaat   
2  Peter Cullen, Sumalee Montano, Frank Welker, J...   
3  Will Friedle, Darren Criss, Constance Zimmer, ...   
4  Nesta Cooper, Kate Walsh, John Michael Higgins...   

                                       ulke     eklenme_tarihi  \
0  United States, I

pandas ile oluşturulan veri çerçevelerini NumPy dizisine dönüştürmek de olanaklıdır. Örneğin:

In [14]:
# Dosyamızın ismini belirleyelim. 
dosya3 = 'netflix.csv'

# veri dosyası adında olarak yükleyelim. Verinin sadece ilk beş satırını alsın ve başlık satırı olmasın.
veri = pd.read_csv(dosya3,nrows = 5, header = None)

# Bu veri çerçevesinden bir NumPy dizisi oluşturalım. np.array() fonksiyonu bu amaç için kullanılmaktadır.
veri_dizisi = np.array(veri)

# veri_dizisinin veri türünü ekrana verelim. 
print(type(veri_dizisi))

[['id' 'tur' 'baslik' 'yonetmen' 'cast' 'ulke' 'eklenme_tarihi'
  'gosterim_tarihi' 'reyting' 'sure' 'listelenme' 'aciklama']
 ['81145628' 'Movie' 'Norm of the North: King Sized Adventure'
  'Richard Finn, Tim Maltby'
  'Alan Marriott, Andrew Toth, Brian Dobson, Cole Howard, Jennifer Cameron, Jonathan Holmes, Lee Tockar, Lisa Durupt, Maya Kay, Michael Dobson'
  'United States, India, South Korea, China' 'September 9, 2019' '2019'
  'TV-PG' '90 min' 'Children & Family Movies, Comedies'
  'Before planning an awesome wedding for his grandfather, a polar bear king must take back a stolen artifact from an evil archaeologist first.']
 ['80117401' 'Movie' 'Jandino: Whatever it Takes' nan 'Jandino Asporaat'
  'United Kingdom' 'September 9, 2016' '2016' 'TV-MA' '94 min'
  'Stand-Up Comedy'
  'Jandino Asporaat riffs on the challenges of raising kids and serenades the audience with a rousing rendition of "Sex on Fire" in his comedy show.']
 ['70234439' 'TV Show' 'Transformers Prime' nan
  'Peter 

## 2. Excel Dosyalarını Okuma

Veri bilimi ile uğraşan herkes, Excel hesap tabloları üzerine çalışmak zorundadır. Dünyada halen Excel formatında (.xls ya da .xlsx) veri setleri kullanılmakta ve dağıtılmaktadır. pandas kütüphanesi ile Excel dosyalarını çalışma ortamına yüklemek mümkündür. Bunun için `pd.ExcelFile()` fonksiyonu kullanılır. Dünya Mutluluk Raporu'nun 2019 verilerini barındıran Excel dosyasını (mutluluk_indeksi.xlsx) kullanalım:

In [16]:
# pandas kütüphanesini yükleyelim.
import pandas as pd

# Dosya ismini belirleyelim.
dosya = 'mutluluk_indeksi.xlsx'

# Excel dosyasını yükleyerek xl veri dosyasını oluşturalım.
xl = pd.ExcelFile(dosya)

# Excel dosyası içinde kayıtlı olan hesap tablolarını (ayrı sekmeler halinde saklanır) ekrana verelim.
print(xl.sheet_names)
print(type(xl))

['2019', '2018']
<class 'pandas.io.excel._base.ExcelFile'>


Dosyada ülkelerin mutluluk indeksi verileri hem 2019 hem de 2018 yılları için ayrı ayrı hesap tablosu sekmesinde saklanmaktadır. Bu her iki hesap tablosu için ayrı veri çerçeveleri oluşturabiliriz. Bunun için `.parse()` yöntemi kullanılır.

In [17]:
df1 = xl.parse('2019')
print(df1.head())

   Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland   7769          1340.0          1587.0   
1             2           Denmark   7600          1383.0          1573.0   
2             3            Norway   7554          1488.0          1582.0   
3             4           Iceland   7494          1380.0          1624.0   
4             5       Netherlands   7488          1396.0          1522.0   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                 1028.000                         0.603       0.271   
3                 1026.000                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                

`.parse()` yönteminin içine hesap tablosunun ismini yazabileceğimiz gibi sıra numarasını (indeksini) de yazabiliriz. 2018 yılı için deneyelim.


In [18]:
df2 = xl.parse(1)
print(df2.head())

   Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland   7632          1305.0          1592.0   
1             2            Norway   7594          1456.0          1582.0   
2             3           Denmark   7555          1351.0          1590.0   
3             4           Iceland   7495          1343.0          1644.0   
4             5       Switzerland   7487          1420.0          1549.0   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                

Şimdi ilk hesap tablosunu, `.parse()` yöntemi ile veri çerçevesi haline getirelim. İlk satırı atlayalım ve sütunları Türkçe'ye çevirelim.

In [20]:
df1 = xl.parse(0, skiprows=[0], names=['Genel Sıralama','Ülke', 'Skor', 'Kişi Başına Milli Gelir Puanı', 
                                       'Sosyal Destek Puanı', 'Sağlıklı Yaşam Beklentisi Puanı', 
                                       'Yaşam Tercihlerinde Özgürlük Puanı', 'Cömertlik Puanı', 'Yolsuzluk Algısı Puanı'])
print(df1.head())
print(type(df1))

   Genel Sıralama         Ülke  Skor  Kişi Başına Milli Gelir Puanı  \
0               2      Denmark  7600                         1383.0   
1               3       Norway  7554                         1488.0   
2               4      Iceland  7494                         1380.0   
3               5  Netherlands  7488                         1396.0   
4               6  Switzerland  7480                         1452.0   

   Sosyal Destek Puanı  Sağlıklı Yaşam Beklentisi Puanı  \
0               1573.0                            0.996   
1               1582.0                         1028.000   
2               1624.0                         1026.000   
3               1522.0                            0.999   
4               1526.0                         1052.000   

   Yaşam Tercihlerinde Özgürlük Puanı  Cömertlik Puanı  Yolsuzluk Algısı Puanı  
0                               0.592            0.252                   0.410  
1                               0.603            0.271  

## 3. SAS / Stata Dosyalarını Okuma

SAS (Statistical Analysis System) gelişmiş bir veri analizi programıdır. Stata (Statistics + data) ise daha çok akademik sosyal bilimler araştırmalarında ve ekonometri amaçlı uygulamalarda kullanılan bir veri analizi programıdır. SAS dosyaları .sas7bdat uzantılıdır. Stata dosyaları ise .dta uzantılıdır. pandas kütüphanesi ile bu iki dosya formatını çalışma ortamından okuyabilmek mümkündür. Çalışma ortamımızda olan ve vergilerle ilgili analiz verilerini içeren tax.sas7bdat dosyasını yüklemeyele çalışalım.


In [21]:
import pandas as pd
from sas7bdat import SAS7BDAT
with SAS7BDAT('tax.sas7bdat') as dosya:
    df_sas = dosya.to_data_frame()
    
print(df_sas.head())

   INC88  TAX88  INC89  TAX89
0  9.215  1.643  9.518  2.125
1  2.047  0.413  2.068  0.565
2  9.989  1.752  9.992  2.221
3  8.321  1.408  8.515  1.905
4  4.588  0.838  4.389  0.943


Aynı işlemi Stata (.dat uzantılı) veri seti için de uygulayalım. pd.read_stata() fonksiyonu işimizi görecektir.

In [22]:
import pandas as pd
df_stata= pd.read_stata('tax.dta')
print(df_stata.head())

   inc88  tax88  inc89  tax89
0  9.215  1.643  9.518  2.125
1  2.047  0.413  2.068  0.565
2  9.989  1.752  9.992  2.221
3  8.321  1.408  8.515  1.905
4  4.588  0.838  4.389  0.943


## 4. MATLAB Dosyalarını Okuma

"Matrix Laboratory" olarak adlandırılan ve mühendislik uygulamalarında bir standart hesaplama dili olan MATLAB ticari paket programlama dilinde veriler .mat uzantılı dosyalar olarak saklanır. Örneğin Stanford Üniversitesi'nden araştırmacıların oluşturduğu araba veri setinin eğitim verilerinin .mat uzantılı versiyonunu scipy kütüphanesini kullanarak çalışma ortamımıza yükleyelim. 

In [23]:
import scipy.io
dosya = 'cars_train_annos.mat'
mat = scipy.io.loadmat(dosya)
print(mat)

{'__header__': b'MATLAB 5.0 MAT-file, Platform: GLNXA64, Created on: Sat Dec 14 14:13:07 2013', '__version__': '1.0', '__globals__': [], 'annotations': array([[(array([[39]], dtype=uint8), array([[116]], dtype=uint8), array([[569]], dtype=uint16), array([[375]], dtype=uint16), array([[14]], dtype=uint8), array(['00001.jpg'], dtype='<U9')),
        (array([[36]], dtype=uint8), array([[116]], dtype=uint8), array([[868]], dtype=uint16), array([[587]], dtype=uint16), array([[3]], dtype=uint8), array(['00002.jpg'], dtype='<U9')),
        (array([[85]], dtype=uint8), array([[109]], dtype=uint8), array([[601]], dtype=uint16), array([[381]], dtype=uint16), array([[91]], dtype=uint8), array(['00003.jpg'], dtype='<U9')),
        ...,
        (array([[26]], dtype=uint8), array([[246]], dtype=uint8), array([[660]], dtype=uint16), array([[449]], dtype=uint16), array([[163]], dtype=uint8), array(['08142.jpg'], dtype='<U9')),
        (array([[78]], dtype=uint8), array([[526]], dtype=uint16), array([[

.mat uzantılı dosyanın sözlük formatında (dict) yüklendiğini görüyoruz. Bu sözlüğün anahtarlarını inceleyelim. 

In [24]:
mat.keys()

dict_keys(['__header__', '__version__', '__globals__', 'annotations'])

.mat dosyalarını okumak için `scipy.io.loadmat()` fonksiyonu, yazmak (kaydetmek) için de `scipy.io.savemat()` fonksiyonu kullanılır.

## 5. Web Ortamından Veri Okuma



Web ortamındaki bağlantıları kullanarak da, Jupyter Notebook içerisine veri aktarımı yapmamız mümkündür. 
İstanbul Büyükşehir Belediyesi Açık Veri Portalı'ndan Ekim 2020 trafik verilerini kullanmak isteyelim.  (https://data.ibb.gov.tr/dataset/saatlik-trafik-yogunluk-veri-seti/resource/949d4a3b-91d2-4c56-b82f-4ef081e39c45)
Bu sayfadaki verileri indirmeden, doğrudan bağlantı adresi alarak, veriyi çalışma ortamına eklememiz için `urllib.request` modülünün `urlretrieve` fonksiyonunu kullanabiliriz. 

In [25]:
from urllib.request import urlretrieve

url = 'https://data.ibb.gov.tr/dataset/3ee6d744-5da2-40c8-9cd6-0e3e41f1928f/resource/949d4a3b-91d2-4c56-b82f-4ef081e39c45/download/traffic_density_202010.csv'
    
urlretrieve(url, 'traffic_density_202010.csv') 
    

('traffic_density_202010.csv', <http.client.HTTPMessage at 0x1d701369988>)

In [26]:
import pandas as pd
df = pd.read_csv('traffic_density_202010.csv', sep=',')
print(df.head())

             DATE_TIME  LONGITUDE   LATITUDE GEOHASH  MINIMUM_SPEED  \
0  2020-10-04 01:00:00  28.866577  41.009216  sxk92c              6   
1  2020-10-05 07:00:00  28.471069  41.135559  sxk4pb              9   
2  2020-10-03 21:00:00  29.119263  41.031189  sxk9qt              6   
3  2020-10-04 18:00:00  29.306030  41.113586  sxkcfu             54   
4  2020-10-01 13:00:00  29.262085  40.877380  sxkb9c              6   

   MAXIMUM_SPEED  AVERAGE_SPEED  NUMBER_OF_VEHICLES  
0             68             30                  24  
1             66             39                  19  
2            122             70                 361  
3            162            111                  13  
4            108             48                 253  


.csv dosyalar haricinde .xlsx dosyaları da bağlantıdan çekebilmek mümkündür. Bunun için pd.read_excel fonksiyonunu kullanarak, argüman oalrak bağlantı adresini kullanacağız. İstanbul Büyükşehir Belediyesi Açık Veri Portalı'nda sunulan İtfaiye Konum Verileri excel dosyasını (https://data.ibb.gov.tr/dataset/itfaiye-istasyonlari-konum-bilgileri/resource/c611b9a1-8a1a-44a9-816b-eb6dfcd37c42) kullanalım. 

In [27]:
# pandas kütüphanesini yükleyelim. 
import pandas as pd

# Bağlantı adresini oluşturulalım: url
url = 'https://data.ibb.gov.tr/dataset/75cd7b09-dafb-41fa-90c9-9c7396a58700/resource/c611b9a1-8a1a-44a9-816b-eb6dfcd37c42/download/itfaiye-konum-verisi.xlsx'

# Excel dosyasındaki tüm sayfaları okuyalım: xl
xl = pd.read_excel(url, sheet_name=None)

# Excel dosyasındaki tüm sayfa isimlerini ekrana verelim.
print(xl.keys())


dict_keys(['Metadata_TR', 'istasyonlar', 'müfrezeler'])


In [29]:
# Excel dosyasındaki ikinci çalışma sayfasının ilk beş satırını ekrana verelim (çalışma sayfası ismini kullanarak)
print(xl['istasyonlar'].head())
print(type(xl))

                                  İstasyon Adı Bulunduğu İlçe  \
0                     Adalar İtfaiye İstasyonu         ADALAR   
1  Akpınar Mahallesi Gönüllü İtfaiye İstasyonu     EYÜPSULTAN   
2                Akşemsettin İtfaiye İstasyonu  GAZİOSMANPAŞA   
3  Alacalı Mahallesi Gönüllü İtfaiye İstasyonu           ŞİLE   
4                  Alibeyköy İtfaiye İstasyonu     EYÜPSULTAN   

                 Koordinat  
0  40.87172994,29.13762931  
1  41.27821768,28.80994231  
2     41.091980, 28.917401  
3  41.18315865,29.45650909  
4     41.079838, 28.937221  
<class 'dict'>
